[e2e] Gutenboarding test flow

razvanpapadopol opened the pull request:
#### Changes proposed in this Pull Request
* Cover the basic free site creation with Gutenboarding.
* <s>After breaking /new after merging an apparently unrelated change in #43012 we decided to add Gutenboarding e2e flow to canaries.</s> Added to canaries in #43024

#### Testing instructions
* tests should pass

Fixes part of #41082

Commit:
Export getRandomInt data-helper function

Commit:
Add Gutenboarding steps to existing e2e test suite

simison commented on test/e2e/package.json:
```diff
@@ -49,7 +49,7 @@
 		"node-slack-upload": "1.2.1",
 		"png-itxt": "1.3.0",
-		"prettier": "npm:wp-prettier@1.19.1",
+		"prettier": "npm:wp-prettier@2.0.5-alpha-2",
```
Unrelated — just wanted to flag here so that we don't accidentally get it in here.

andrewserong replied:
I've removed this change during rebase.

andrewserong commented:
Rebased to remove the change to `package.json` and to factor in the changes in https://github.com/Automattic/wp-calypso/pull/41398/ where the `package.lock` file was removed.

andrewserong committed:
Add class for continue button on mobile to StylePreviewPage e2e test class

Commit:
Update Gutenboarding flow e2e tests:
- remove vertical sub-step
- update selectors
- add Plans step

bsessions85 commented on test/e2e/lib/pages/gutenboarding/acquire-intent-page.js:
```diff
@@ -0,0 +1,42 @@
+	async enterSiteTitle( siteTitle ) {
+		const siteTitleSelector = By.css( '#site-title__input' );
+		const siteTitleElement = this.driver.findElement( siteTitleSelector );
+		return await siteTitleElement.sendKeys( siteTitle );
```
using `driverHelper.setWhenSettable()` is usually more stable when entering text.

razvanpapadopol replied:
I remember we were using `sendKeys` to work with `contenteditable` but now that we have an `<input/>` I've switched to `driverHelper.setWhenSettable()`. Thanks for the suggestion!

bsessions85 reviewed (commented):
This is looking good. One small change and fix one test is failing on mobile and you're good.

simison commented:
@razvanpapadopol would this be ready to go after rebase?

simison commented on test/e2e/specs/wp-gutenboarding-spec.js:
```diff
@@ -23,18 +28,46 @@ before( async function () {
+			const designSelectorPage = await DesignSelectorPage.Expect( driver );
+			const designOptionsCount = await designSelectorPage.getDesignOptionsCount();
+			return await designSelectorPage.selectDesign(
+				dataHelper.getRandomInt( 1, designOptionsCount )
```
We might wanna make sure to intentionally pick a non-premium design.

Could be either by picking the first one, or ensuring there's no `.design-selector__premium-badge` within the design.

Commit:
- Update existing Gutenboarding e2e test flow.
- Add new Gutenboarding test to cover skipping first step and selecting paid design.

Commit:
- Fix tests on mobile.
- Cleanup.

simison commented on client/landing/gutenboarding/onboarding-block/design-selector/index.tsx:
```diff
@@ -59,6 +59,7 @@ const DesignSelector: React.FunctionComponent = () => {
 						<button
 							key={ design.slug }
 							className="design-selector__design-option"
+							data-e2e-button={ design.is_premium ? 'paidOption' : 'freeOption' }
```
I like how this is clear and results in more stable tests 👍

simison commented on test/e2e/lib/pages/gutenboarding/acquire-intent-page.js:
```diff
@@ -0,0 +1,35 @@
+	async skipStep() {
+		const skipButtonSelector = By.css(
+			this.screenSize === 'MOBILE'
+				? '.acquire-intent__footer .acquire-intent__skip-site-title'
```
`acquire-intent__footer` still there?

razvanpapadopol replied:
Yes, it's just in a different position. 
We need separate HTML for desktop/mobile because of the layout: skip button on desktop is relative to the input field. I feel this would change soon so we'll update also the selector here 🤓

simison commented on test/e2e/specs/wp-gutenboarding-spec.js:
```diff
@@ -26,24 +31,73 @@ before( async function () {
+			assert( blockExists, 'Onboarding block is not rendered' );
+		} );
+
+		step( 'Can see Acquire Intent and set a vertical and site title', async function () {
```
"set a vertical" can be removed

simison reviewed (approved):
Overall looking great! and CI is green! :tada:

Left some notes about "intent capture" terms still here but not a blocker.

simison commented:
Please rebase just before merging to ensure we're not missing any breaking changes in the e2e-flow from `master`.

Commit:
Update step description in Gutenboarding e2e test.
